Add Auto Discovery Support For Sagemaker

charleschangdp opened the pull request:
Sagemaker has several tools that generate metrics under different namepsaces. Working based on the AWS [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/monitoring-cloudwatch.html), adding the following:

- AWS/SageMaker - model invocation and loading metrics 
- /aws/sagemaker/Endpoints - Sagemaker Endpoint health metrics
- /aws/sagemaker/TrainingJobs - Training job metrics
- /aws/sagemaker/ProcessingJobs - Processing job metrics
- /aws/sagemaker/TransformJobs - Batch transform job metrics
- /aws/sagemaker/InferenceRecommendationsJobs - Inference Recommender metrics
- AWS/Sagemaker/ModelBuildingPipeline - Model pipeline metrics

cristiangreco reviewed (commented):
Thanks for the contribution @charleschangdp!

Do you have an example of the metrics being generated? I'd expect we would need to add `DimensionRegexps` in the services definition, why are those not needed?

charleschangdp commented:
Here's an example from Prometheus during my local testing. (Account number and endpoint name redacted):

```
aws_sagemaker_endpoints_cpuutilization_sum{account_id="000000000000", dimension_EndpointName="example-endpoint", dimension_VariantName="AllTraffic", instance="yace:8080", job="yace", name="global", region="us-west-2"}
0.0883032
```

In this specific example, it looks like the dimensions(VariantName and EndpointName) are discovered automatically. And there aren't additional information in the ARN (`arn:aws:sagemaker:us-west-2:000000000000:endpoint/example-endpoint` ) to create additional dimensions.

I've checked all 7 namespaces and I've yet to find any namespaces that warrant additional dimensions.

cristiangreco commented:
@charleschangdp sorry for the late reply! Thanks for the clarification, I've got another request: would you mind adding a test in the `pkg/job/maxdimassociator` package for SageMaker resources? You can look at [this PR](https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/937/files) for an example of what needs to be tested.

charleschangdp commented:
@cristiangreco: Thanks for sharing the example of the tests. After reviewing the examples and thinking about it some more, I've added regexp to the following:
- AWS/SageMaker - model invocation and loading metrics
- /aws/sagemaker/Endpoints - Sagemaker Endpoint health metrics
- AWS/Sagemaker/ModelBuildingPipeline - Model pipeline metrics
- /aws/sagemaker/InferenceRecommendationsJobs - Inference Recommender metrics

But for the others I've listed below, only part of the dimension value is contained in the ARN. (see [documentionation](https://docs.aws.amazon.com/sagemaker/latest/dg/monitoring-cloudwatch.html) under section **Dimensions for Processing Job, Training Job and Batch Transform Job Instance Metrics**) Thus I can't figure out how to add regexp.
- /aws/sagemaker/TrainingJobs - Training job metrics
- /aws/sagemaker/ProcessingJobs - Processing job metrics
- /aws/sagemaker/TransformJobs - Batch transform job metrics

I've also added more specific resource filters for all the above.

cristiangreco commented:
@charleschangdp thanks for the great contribution! We might merge the PR as-is and it's already of great value.

Regarding your last comment:

> But for the others I've listed below, only part of the dimension value is contained in the ARN. (see [documentionation](https://docs.aws.amazon.com/sagemaker/latest/dg/monitoring-cloudwatch.html) under section Dimensions for Processing Job, Training Job and Batch Transform Job Instance Metrics) Thus I can't figure out how to add regexp.

I found this documentation:

<img width="699" alt="Screenshot 2023-06-14 at 11 50 00" src="https://github.com/nerdswords/yet-another-cloudwatch-exporter/assets/316923/c693c10b-96f2-4e43-a218-4f5d976abc25">

Not the first time we see something similar in CloudWatch, we've already added a bit of custom logic for `AmazonMQ` in the new maxdimassociator, see [here](https://github.com/nerdswords/yet-another-cloudwatch-exporter/blob/c7807a770bb427f8ddb2c7becac51185fb3e8230/pkg/job/maxdimassociator/associator.go#L147-L152). 

If you're using the various `{Processing,Training,Transform}Jobs` in your account (I don't have any of those SageMaker resources setup) and have got some time to work on it, would be great to write another bit of custom logic that parses the `Host` dimension and captures the job name. What do you think? Otherwise, I'm happy to merge the PR as-is for now.

charleschangdp commented:
@cristiangreco: I looked into the AmazonMQ example you provided and it made sense. I tried my hand at adding training-job. The `maxdimassociator`  test case worked. When I tried end-to-end testing with docker-compose, it didn't behave as expected. After some debugging, I noticed that when I run docker-compose, the import `discovery.go` is somehow not resolving to the changes I've made in `associator.go` buildLabelsMap function. 

This is my first rodeo with Golang, so it will take some time for me to figure this out. Can you merge the current PR (hopefully I've fixed the all linting issue)? I'll share with you a separate branch with the proposed custom logic. Maybe you'll see some issue that I didn't.

charleschangdp commented:
Here's a branch I've created with my attempt to add dimenionregexps for Sagemaker training jobs. https://github.com/charleschangdp/yet-another-cloudwatch-exporter/tree/sagemaker_custom_logic

I've made changes to:

- test case: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/job/maxdimassociator/associator_sagemaker_training_test.go#L42
- services.go: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/config/services.go#L755
- associator.go: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/job/maxdimassociator/associator.go#L153
- yace-config.yml: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/sagemaker_custom_logic/docker-compose/yace-config.yaml

After running `docker-compose`, I get this message:

> {"account":"0000000000","arn":"","job_type":"sagemaker-training","level":"info","msg":"No metrics data found","region":"us-west-2","time":"2023-06-15T21:25:23Z"}


After some debugging, I noticed that discover.go, for some reason I can't figure out right now, is not resolving to the changes I've made in associator.go. Thus the code is still looking to match the whole value in Host dimension (ex: `example-training-job/algo-1`) to an ARN (ex: `arn:aws:sagemaker:us-west-2:000000000000:training-job/example-training-job` ) I was expecting the dimension value to have been truncated. (ex: `example-training-job`)

thomaspeitz commented:
I will merge this one, as it is already a nice addition.

Feel free to add / improve more in another PR :)

Want to have this already in! Thanks for the nice contribution! <3

cristiangreco commented:
> Here's a branch I've created with my attempt to add dimenionregexps for Sagemaker training jobs. https://github.com/charleschangdp/yet-another-cloudwatch-exporter/tree/sagemaker_custom_logic
> 
> I've made changes to:
> 
>     * test case: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/job/maxdimassociator/associator_sagemaker_training_test.go#L42
> 
>     * services.go: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/config/services.go#L755
> 
>     * associator.go: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/1f2a6c98eb3cb4c44564e1cee2c6d84dbde75c00/pkg/job/maxdimassociator/associator.go#L153
> 
>     * yace-config.yml: https://github.com/charleschangdp/yet-another-cloudwatch-exporter/blob/sagemaker_custom_logic/docker-compose/yace-config.yaml
> 
> 
> After running `docker-compose`, I get this message:
> 
> > {"account":"0000000000","arn":"","job_type":"sagemaker-training","level":"info","msg":"No metrics data found","region":"us-west-2","time":"2023-06-15T21:25:23Z"}
> 
> After some debugging, I noticed that discover.go, for some reason I can't figure out right now, is not resolving to the changes I've made in associator.go. Thus the code is still looking to match the whole value in Host dimension (ex: `example-training-job/algo-1`) to an ARN (ex: `arn:aws:sagemaker:us-west-2:000000000000:training-job/example-training-job` ) I was expecting the dimension value to have been truncated. (ex: `example-training-job`)

Sorry for the late reply! As Thomas said, the work you've done so far is great and worth merging! Please go ahead and open another PR for the changes above, I'll try to help you test them.
